packages/amplify_datastore/test/native_amplify_api_test.dart

```diff
@@ -0,0 +1,993 @@
+
+  tearDown(() async {
+    await Amplify.reset();
+  });
+
+  group('NativeAmplifyAPI', () {
+    group('Query', () {
+      test('Should handle empty request/response', () async {
+        String document = '';
+        Map<String, dynamic> emptyMap = {};
+        String payloadJson = '{"data":{},"errors":[]}';
+
+        mockAPIPlugin.queryMethod = <String>(GraphQLRequest mockRequest) {
+          expect(mockRequest.document, document);
+          expect(mockRequest.variables, emptyMap);
+          expect(mockRequest.apiName, null);
+          expect(mockRequest.authorizationMode, null);
+          expect(mockRequest.headers, null);
+          expect(mockRequest.decodePath, null);
+          expect(mockRequest.modelType, null);
+
+          return GraphQLOperation<String>(
+            CancelableOperation<GraphQLResponse<String>>.fromValue(
+              GraphQLResponse<String>(
+                data: null as String?,
+                errors: [],
+              ),
+            ),
+          );
+        };
+
+        NativeAmplifyApi nativeAmplifyApi = NativeAmplifyApi({});
+
+        NativeGraphQLRequest request = NativeGraphQLRequest(document: '');
+        NativeGraphQLResponse response = await nativeAmplifyApi.query(request);
+
+        expect(response.payloadJson, payloadJson);
+      });
+
+      test('Should handle filled request/response', () async {
+        String document = 'document';
+        String apiName = 'apiName';
+        String variable1Key = 'variable1';
+        int variable1Value = 1;
+        String variable2Key = 'variable2';
+        String variable2Value = 'test string';
+        String variablesJson =
+            '{"$variable1Key": $variable1Value, "$variable2Key": "$variable2Value"}';
+        String responseType = 'responseType';
+        String decodePath = 'decodePath';
+        String options = 'options';
+        String data = '{"data1":1,"data2":"data string"}';
+        String errorMessage1 = 'errorsJson1';
+        String errorMessage2 = 'errorsJson2';
+        String payloadJson =
+            '{"data":$data,"errors":[{"message":"$errorMessage1"},{"message":"$errorMessage2"}]}';
+        String authMode = 'apiKey';
+
+        mockAPIPlugin.queryMethod = <String>(GraphQLRequest mockRequest) {
+          expect(mockRequest.id.length, greaterThan(0));
+          expect(mockRequest.document, document);
+          expect(mockRequest.variables[variable1Key], variable1Value);
+          expect(mockRequest.variables[variable2Key], variable2Value);
+          expect(mockRequest.apiName, apiName);
+          expect(mockRequest.authorizationMode, APIAuthorizationType.apiKey);
+          expect(mockRequest.headers, null);
+          expect(mockRequest.decodePath, null);
+          expect(mockRequest.modelType, null);
+
+          return GraphQLOperation<String>(
+            CancelableOperation<GraphQLResponse<String>>.fromValue(
+              GraphQLResponse<String>(
+                data: data as String?,
```

Comment:
Note: The fact that this is making you cast to `String?` is odd. I think this is a bug in Dart. I created a minimal repro of this and opened an issue: https://github.com/dart-lang/sdk/issues/55958